presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpGenericPattern.java

```diff
@@ -0,0 +1,435 @@
+                }
+                else { // case 2 in the above comment
+                    backref = nextByte - '0';
+                    if (backref < 0 || backref > 9) { // not using checkArgument because `.toStringUtf8` is expensive
+                        throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: " + replacement.toStringUtf8());
+                    }
+                    if (region.numRegs <= backref) {
+                        throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: unknown group " + backref);
+                    }
+                    idx++;
+                    while (idx < replacement.length()) { // Adaptive group number: find largest group num that is not greater than actual number of groups
+                        int nextDigit = replacement.getByte(idx) - '0';
+                        if (nextDigit < 0 || nextDigit > 9) {
+                            break;
+                        }
+                        int newBackref = (backref * 10) + nextDigit;
+                        if (region.numRegs <= newBackref) {
+                            break;
+                        }
+                        backref = newBackref;
+                        idx++;
+                    }
+                }
+                int beg = region.beg[backref];
+                int end = region.end[backref];
+                if (beg != -1 && end != -1) { // the specific group doesn't exist in the current match, skip
+                    result.appendBytes(source.slice(beg, end - beg));
+                }
+            }
+            else { // case 3 and 4 in the above comment
+                if (nextByte == '\\') {
+                    idx++;
+                    if (idx == replacement.length()) { // not using checkArgument because `.toStringUtf8` is expensive
+                        throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: " + replacement.toStringUtf8());
+                    }
+                    nextByte = replacement.getByte(idx);
+                }
+                result.appendByte(nextByte);
+                idx++;
+            }
+        }
+    }
+
+    public Block extractAll(Slice source, long groupIndex)
+    {
+        if (selectedRegexLibrary == RegexLibrary.RE2J) {
+            String inputString =  source.toStringUtf8();
+            com.google.re2j.Matcher matcher = re2jPattern.matcher(inputString);
+            validateRe2jGroup(Ints.checkedCast(groupIndex), matcher.groupCount());
+
+            BlockBuilder blockBuilder = VARCHAR.createBlockBuilder(new BlockBuilderStatus(), 32);
+            while (true) {
+                 if (!matcher.find()) {
+                    break;
+                 }
+
+                 if (groupIndex == 0) { //does not need any specific group, return the whole match
+                     Slice slice = source.slice(matcher.start(), matcher.end() - matcher.start());
```

Comment:
As this might happen for non-ASCII chars, for now it should be ok. If there is more interest in searching UTF8 chars then we can verify this functionality later.